### PR TITLE
Use property type for data model instances

### PR DESCRIPTION
**Pull request by emelvarol-cognite**

### CODE REVIEW

File: src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala [hunk 1]
```diff
@@ -74,44 +72,43 @@ class DataModelInstanceRelation(config: RelationConfig, modelExternalId: String)
       dmi: DataModelInstanceQueryResponse,
       requiredPropsArray: Seq[String]): ProjectedDataModelInstance = {
     val dmiProperties = dmi.properties.getOrElse(Map())
-    ProjectedDataModelInstance(
+    val res = ProjectedDataModelInstance(
```
Comment (hakontro) [inline src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala:75]: I don't think you need the `res` thing here, perhaps it's leftovers from development?

File: src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala [hunk 2]
```diff
@@ -74,44 +72,43 @@ class DataModelInstanceRelation(config: RelationConfig, modelExternalId: String)
...
       externalId = dmi.properties
         .flatMap(_.get("externalId"))
-        .flatMap(_.asString)
+        .map(_.asInstanceOf[StringProperty].value)
         .getOrElse(
           throw new CdfSparkException("Can't read data model instance, `externalId` is missing.")),
       properties = requiredPropsArray.map { name: String =>
```
Comment (hakontro) [inline src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala:81]: Are these properties required to exist in the result, as null or some, or are they actually required to be populated? Seems this code accepts null entries for the required properties, which was a little puzzling to me.

File: src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala [hunk 3]
```diff
@@ -197,81 +194,79 @@ class DataModelInstanceRelation(config: RelationConfig, modelExternalId: String)
           (index, field.name, propertyType)
       }
 
-    def parseRow(row: Row): DataModelInstance = {
+    def parseRow(row: Row): DataModelInstanceCreate = {
```
Comment (hakontro) [inline src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala:197]: If we curried it like `parseRow(row: Row)(schema: StructType)`, we could avoid the function in the function. `parseRow` now uses parameters that are in the outer function `fromRow`, and I think it's nicer to be able to have it all in one place!

File: src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala [hunk 4]
```diff
@@ -197,81 +194,79 @@ class DataModelInstanceRelation(config: RelationConfig, modelExternalId: String)
...
-            row.get(index),
-            (_: Any) =>
-              throw SparkSchemaHelperRuntime
-                .badRowError(row, name, propT._1, "")
-          )
-      }.toMap
-      DataModelInstance(modelExternalId, properties = Some(propertyValues))
+      val propertyValues: Map[String, PropertyType] = properties
+        .map {
+          case (index, name, propT) =>
```
Comment (hakontro) [inline src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala:204]: Could some of this code be nicer if we made a case class for `propertyTypes`? Maybe there's already one we can use!
`val propertyTypes: Map[String, (String, Boolean)]`

File: src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala [hunk 5]
```diff
@@ -197,81 +194,79 @@ class DataModelInstanceRelation(config: RelationConfig, modelExternalId: String)
...
+    case x: Array[Double] => ArrayProperty(x.toVector.map(Float64Property))
+    case x: Array[Int] => ArrayProperty(x.toVector.map(Int32Property))
+    case x: Array[Float] => ArrayProperty(x.toVector.map(Float32Property))
+    case x: Array[Long] => ArrayProperty(x.toVector.map(Int64Property))
+    case x: Array[String] => ArrayProperty(x.toVector.map(StringProperty))
+    case x: Array[Boolean] => ArrayProperty(x.toVector.map(BooleanProperty))
+    case x: Array[java.math.BigDecimal] =>
+      ArrayProperty(x.toVector.map(i => Float64Property(i.doubleValue())))
+    case x: Array[java.math.BigInteger] =>
+      ArrayProperty(x.toVector.map(i => Int64Property(i.longValue())))
```
Comment (hakontro) [inline src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala:257]: Let's add a final `case: _` just in case something crazy happens.

File: src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala [hunk 6]
```diff
@@ -197,81 +194,79 @@ class DataModelInstanceRelation(config: RelationConfig, modelExternalId: String)
...
     parseRow
   }
 }
 
 object DataModelInstanceRelation {
-  private def propertyNotNullableMessage(propertyType: String) =
-    s"Property of $propertyType type is not nullable."
   private def unknownPropertyTypeMessage(a: Any) = s"Unknown property type $a."
 
   private def notValidPropertyTypeMessage(a: Any, propertyType: String) =
```
Comment (hakontro) [inline src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala:232]: This seems to be thrown when someone tries to write something that can't be parsed to the property type of the schema. We accept multiple types of numbers casted to Float32 for example, perhaps there's a more informative way to let the user know what they need to do!

**Review by hakontro — COMMENTED** [review]
Very nice, had some comments.

File: src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala [hunk 7]
```diff
@@ -291,92 +286,103 @@ object DataModelInstanceRelation {
...
-        case x: java.math.BigDecimal => jsonFromDouble(x.doubleValue)
-        case x: java.math.BigInteger => jsonFromDouble(x.doubleValue)
+      case "float32" => {
+        case x: Float => Float32Property(x)
+        case x: Double => Float32Property(x.toFloat)
+        case x: Int => Float32Property(x.toFloat)
+        case x: Long => Float32Property(x.toFloat)
+        case x: java.math.BigDecimal => Float32Property(x.floatValue())
+        case x: java.math.BigInteger => Float32Property(x.floatValue())
+        case a => throw new CdfSparkException(notValidPropertyTypeMessage(a, propertyType))
```
Comment (vuhainguyen) [inline src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala:298]: I think handling each case in a separate function will reduce the cyclomatic complexity and make the code easier to read hehe 😁

File: src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala (hunk 2 above)
Comment (emelvarol-cognite) [reply src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala:81]: required means, the selected properties i believe, name is just deceiving.

File: src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala (hunk 1 above)
Comment (emelvarol-cognite) [reply src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala:75]: yis

File: src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala (hunk 3 above)
Comment (emelvarol-cognite) [reply src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala:197]: The beauty of this version is that the part in `fromRow` is done once and parseRow is called for each row. I'll try moving the first part out to another function.

Comment (hakontro) [reply src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala:197]: That _is_ nice.. but I think we get the same benefit from currying.. Do you know @wjoel ?

File: src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala [hunk 8]
```diff
@@ -291,92 +286,103 @@ object DataModelInstanceRelation {
...
-        case x: Iterable[_] if x.isEmpty => Json.arr()
+        case x: Iterable[_] if x.isEmpty => ArrayProperty(Vector.empty)
         case x: Iterable[Boolean] @unchecked if x.head.isInstanceOf[Boolean] =>
-          Json.fromValues(x.map(Json.fromBoolean))
+          ArrayProperty(x.toVector.map(BooleanProperty))
+        case a => throw new CdfSparkException(notValidPropertyTypeMessage(a, propertyType))
+      }
+      case "int[]" | "int32[]" => {
+        case x: Iterable[_] if x.isEmpty => ArrayProperty(Vector.empty)
+        case x: Iterable[Int] @unchecked if x.head.isInstanceOf[Int] =>
```
Comment (wjoel) [inline src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala:372]: I think you can just do `case x: Iterable[_] if x.head.isInstanceOf[Int]` here.

The `@unchecked` is needed because due to type erasure it's not truly possible to match on `Iterable[Int]`, at runtime it's only `Iterable` with no type information stored for the element types, so the type check against `Iterable[Int]` is a lie, and that's why the compiler requires us to mark it as `@unchecked` (as it can't be checked).

File: src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala [hunk 9]
```diff
@@ -291,92 +286,103 @@ object DataModelInstanceRelation {
...
+          ArrayProperty(x.toVector.map(Int32Property))
+        case x: Iterable[java.math.BigInteger] @unchecked if x.head.isInstanceOf[java.math.BigInteger] =>
+          ArrayProperty(x.toVector.map(i => Int32Property(i.intValue())))
```
Comment (wjoel) [inline src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala:375]: I'm not sure we should allow this. Maybe we have to because `BigInteger` is used often by Spark, even if the values are small? But it will not work well if the value is larger than `Integer.MaxValue`...

File: src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala (hunk 3 above)
Comment (emelvarol-cognite) [reply src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala:197]: I thought it is just a way of grouping the arguments? 🤔

Comment (hakontro) [reply src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala:197]: I may not be using the term right, but essentially having multiple argument lists to the function, so that we can do
`def parseRow(row: Row)(schema: StructType): DataModelInstanceCreate`
and then use it like
`parseRow(someSchema)`, which will then return a function of the type `Row => DataModelInstanceCreate`..

Seems the actual definition is https://www.baeldung.com/scala/currying#currying

> Currying is the process of converting a function with multiple arguments into a sequence of functions that take one argument. Each function returns another function that consumes the following argument.

File: src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala (hunk 4 above)
Comment (emelvarol-cognite) [reply src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala:204]: yes sure, we have a type from scala-sdk which is nice

File: src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala (hunk 3 above)
Comment (emelvarol-cognite) [reply src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala:197]: I thought you wanted me to remove the inner function and make it one, sorry I misunderstood

Comment (emelvarol-cognite) [reply src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala:197]: we don't make use of `schema: StructType` directly in `parseRow`. It uses an array calculated in `fromRow` using the schema. Should I take that one as input?

File: src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala (hunk 9 above)
Comment (wjoel) [reply src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala:364]: I don't know how often they end up being inferred, so I'm a bit unsure, but I guess we can remove it for now, and add it back if we find out it's a common thing?

File: src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala (hunk 8 above)
Comment (wjoel) [reply src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala:361]: It's not about IntelliJ's type matching, it's about the JVM and type erasure.
At runtime, the type of `x` will be `Iterable`. It's not `Iterable[Int]`, `Iterable[Double]`, etc. it's compiled into only `Iterable`. Type parameters don't exist in JVM bytecode.

The Scala compiler knows this, and it knows that it can't actually check if `x` is of type `Iterable[Int]`. That's why it fails to compile. Adding `@unchecked` says "trust me, I know what I'm doing" and in our case we know what we're doing (we think) because of the `x.head.isInstanceOf[Int]` check.

File: src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala (hunk 9 above)
Comment (emelvarol-cognite) [reply src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala:364]: I already hit class cast exception in the tests when I removed `BigDecimal` from `float32` and `BigInteger` from `int`. So I keep it if ok.

File: src/test/scala/cognite/spark/v1/DataModelInstancesRelationTest.scala [hunk 10]
```diff
@@ -39,16 +39,13 @@ class DataModelInstancesRelationTest
...
 
   private def byExternalId(modelExternalId: String, externalId: String): String =
     listInstances(
       modelExternalId,
-      filter = Some(DMIEqualsFilter(Seq("instance", "externalId"), Json.fromString(externalId))))
-      .flatMap(_.properties.flatMap(_.get("externalId")).toList)
-      .flatMap(_.asString.toList)
-      .head
+      filter = Some(DMIEqualsFilter(Seq("instance", "externalId"), StringProperty(externalId))))
+      .flatMap(_.properties.flatMap(_.get("externalId")).toList).head.asInstanceOf[StringProperty].value
```
[the remark this answers is not in the document]
Comment (vuhainguyen) [reply src/test/scala/cognite/spark/v1/DataModelInstancesRelationTest.scala:48]: Suggestion:
```scala
    listInstances(
      modelExternalId,
      filter = Some(DMIEqualsFilter(Seq("instance", "externalId"), StringProperty(externalId))))
      .flatMap(_.properties.flatMap(_.get("externalId")))
      .collectFirst { case StringProperty(id) => id }
      .getOrElse("")
```
If you don't want to have getOrElse at the end then return `Option[String]` and later do this when comparing
```scala
byExternalId(primitiveExtId, randomId) shouldBe Some(randomId)
```

File: src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala [hunk 11]
```diff
@@ -77,14 +63,12 @@ class DataModelInstanceRelation(config: RelationConfig, modelExternalId: String)
     ProjectedDataModelInstance(
       externalId = dmi.properties
         .flatMap(_.get("externalId"))
-        .flatMap(_.asString)
+        .map(_.asInstanceOf[StringProperty].value)
```
Comment (vuhainguyen) [inline src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala:66]: I think you can also use `collect { case StringProperty(externalId) => externalId }` here instead

File: src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala [hunk 12]
```diff
@@ -93,25 +77,25 @@ class DataModelInstanceRelation(config: RelationConfig, modelExternalId: String)
...
+        Some(DMIRangeFilter(Seq(modelExternalId, attribute), gt = Some(parsePropertyValue(value))))
       case LessThanOrEqual(attribute, value) =>
-        Some(DMIRangeFilter(Seq(modelExternalId, attribute), lte = Some(parseValue(value))))
+        Some(DMIRangeFilter(Seq(modelExternalId, attribute), lte = Some(parsePropertyValue(value))))
       case LessThan(attribute, value) =>
-        Some(DMIRangeFilter(Seq(modelExternalId, attribute), lt = Some(parseValue(value))))
+        Some(DMIRangeFilter(Seq(modelExternalId, attribute), lt = Some(parsePropertyValue(value))))
       case And(f1, f2) =>
         (getInstanceFilter(f1) ++ getInstanceFilter(f2)).reduceLeftOption((sf1, sf2) =>
           DMIAndFilter(Seq(sf1, sf2)))
```
Comment (vuhainguyen) [inline src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala:101]: I have a doubt about the case Or:
```scala
case Or(f1, f2) =>
    (getInstanceFilter(f1), getInstanceFilter(f2)) match {
      case (Some(sf1), Some(sf2)) =>
        Some(DMIOrFilter(Seq(sf1, sf2)))
      case _ =>
        None
    }
```
Don't we want  to have 2 case below as well ? 
```scala
case (Some(sf1), None) =>
            Some(DMIOrFilter(Seq(sf1)))
case (None, Some(sf2)) =>
            Some(DMIOrFilter(Seq(sf2)))
```

File: src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala [hunk 13]
```diff
@@ -179,99 +163,93 @@ class DataModelInstanceRelation(config: RelationConfig, modelExternalId: String)
...
+                Some(toPropertyType(propT.`type`)(row.get(index)))
+            })
+        }
+        .collect { case (a, Some(value)) => a -> value }
+        .toMap
+
+      DataModelInstanceCreate(modelExternalId, properties = Some(propertyValues))
     }
-    parseRow
+    parseRow(indexedPropertyList)
```
Comment (emelvarol-cognite) [inline src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala:207]: @hakontro does this look better, can i resolve this one?

File: src/test/scala/cognite/spark/v1/DataModelInstancesRelationTest.scala (hunk 10 above)
[the remark this answers is not in the document]
Comment (emelvarol-cognite) [reply src/test/scala/cognite/spark/v1/DataModelInstancesRelationTest.scala:48]: I thought here if the code and the API works correctly, we would always have a head so worst case the test fail as it should. But I'll fix it 👍

File: src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala (hunk 12 above)
Comment (emelvarol-cognite) [reply src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala:101]: I had a reason for this, let me try recalling...

Comment (emelvarol-cognite) [reply src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala:101]: So the reason we get `None` here, is we can't match the filter any case here. So we should not push down by only one condition we can match to something. It may be a wider filter when we `or` and we pushdown more than we should. So we let spark filter the whole thing in that case, not to miss requested data.

File: src/test/scala/cognite/spark/v1/DataModelInstancesRelationTest.scala (hunk 10 above)
[the remark this answers is not in the document]
Comment (wjoel) [reply src/test/scala/cognite/spark/v1/DataModelInstancesRelationTest.scala:48]: It should return `Option[String]` or throw an exception in `getOrElse` with a message explaining what went wrong (if it's unexpected and we can't recover). Converting to an empty string is no good.

[the remark this answers is not in the document]
Comment (emelvarol-cognite) [reply src/test/scala/cognite/spark/v1/DataModelInstancesRelationTest.scala]: took @vuhainguyen 's suggestion without `getOrElse`

**Comment by emelvarol-cognite** [comment]
> I found a typo, and there's uncommented stuff that needs to be cleaned up, but overall this looks really good now.

Thanks a lot for reviewing again! Can you point me to the typo pls? I'll go through the pr to find commented stuff and clean up. Also, I want to add a test for reads on new types I added. Then I'll ping you again for a review again 🙏

File: src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala [hunk 14]
```diff
@@ -291,98 +287,169 @@ object DataModelInstanceRelation {
...
+  }
+
+  private def toTimestampProperty: Any => TimeStampProperty = {
+    case x: Instant =>
+      TimeStampProperty(OffsetDateTime.ofInstant(x, ZoneId.systemDefault()).toZonedDateTime)
+    case x: java.sql.Timestamp =>
+      // Using ZonedDateTime directly without OffSetDateTime, adds ZoneId to the end of encoded string
+      // 2019-10-02T07:00+09:00[Asia/Seoul] instead of 2019-10-02T07:00+09:00, API does not like it.
+      TimeStampProperty(OffsetDateTime.ofInstant(x.toInstant, ZoneId.systemDefault()).toZonedDateTime)
+    case a => throw new CdfSparkException(notValidPropertyTypeMessage(a, "timestamp"))
```
[the remark this answers is not in the document]
Comment (wjoel) [reply src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala:304]: It is nice, but we shouldn't use the system default `ZoneId`, rather convert it to UTC explicitly.
Maybe `LocalDateTime.from(ts.toLocalDateTime.atZone(ZoneOffset.UTC))` or similar. Tests would be needed to make sure we don't get it wrong for different time zones.

[the remark this answers is not in the document]
Comment (emelvarol-cognite) [reply src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala:304]: `TimeStampProperty` wants `ZonedDateTime` as value type. I'll see if it can be adapted to that.

File: src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala [hunk 15]
```diff
@@ -179,99 +165,104 @@ class DataModelInstanceRelation(config: RelationConfig, modelExternalId: String)
...
+      ArrayProperty(x.toVector.map(i => Float64Property(i.doubleValue())))
+    case x: Array[java.math.BigInteger] =>
+      ArrayProperty(x.toVector.map(i => Int64Property(i.longValue())))
+    case x: LocalDate => DateProperty(x)
+    case x: java.sql.Date => DateProperty(x.toLocalDate)
+    case x: LocalDateTime => DateProperty(x.toLocalDate)
+    case x: Instant => TimeStampProperty(x.atZone(ZoneId.systemDefault()))
+    case x: java.sql.Timestamp =>
+      TimeStampProperty(x.toInstant.atZone(ZoneId.systemDefault()))
+    case x => throw new CdfSparkException(s"Cannot parse the value with udentified type: $x")
```
Comment (wjoel) [inline src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala:248]: Here's the typo, in "udentified". I don't know what happened to my previous comment... but my suggestion was `s"Unsupported value ${x.toString} of type ${x.getClass.getName}"`

File: src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala [hunk 16]
```diff
@@ -179,99 +165,104 @@ class DataModelInstanceRelation(config: RelationConfig, modelExternalId: String)
...
+    case Float64Property(value) => value
+    case BooleanProperty(value) => value
+    case StringProperty(value) => value
+    case DateProperty(value) => value
+    case TimeStampProperty(value) => value
+    case DirectRelationProperty(value) => value
+    case GeographyProperty(value) => value
+    case GeometryProperty(value) => value
+    case ArrayProperty(values) => values.map(fromPropertyType)
+    case x => throw new CdfSparkException(s"Unknown property type with value $x")
```
Comment (wjoel) [inline src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala:264]: Similarly, `s"Unknown property type ${x.getClass.getName} with value ${x.toString}"` could be more helpful.

File: src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala (hunk 14 above)
[the remark this answers is not in the document]
Comment (wjoel) [reply src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala:304]: Well, this PR is quite big already, so it might be good to get this merged first even with a simple-but-not-always-correct solution for timestamps first, and then add tests and fix edge cases in another (shorter) PR.

[the remark this answers is not in the document]
Comment (emelvarol-cognite) [reply src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala]: Sure I think this version can be merged 👍 I can create another PR to test type conversions and filters on timestamp etc.

**Comment by emelvarol-cognite** [comment]
Can I get one last look? I believe it is clean now. I'll add tests for timestamp conversions and check filters on new types as a separate PR as it is already pretty big (as @wjoel suggested)

**Review by wjoel — APPROVED** [review]
LGTM, nice work!

**Comment by wjoel** [comment]
For the next few PRs it would be great if we could get the test coverage up to what it was, -1.24% is significant and we should have tests that cover all types, eventually.
